design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,124 @@
+a host to the cluster boils down to powering it on with a configuration ISO provided.
+
+### Goals
+
+Expose the Ironic API to attach non-bootable iso images via Metal3.
+
+### Non-Goals
+
+Supporting non-ISO images i.e. media types other than CD/DVD, supporting
+drivers other than Redfish and its derived drivers.
+
+## Proposal
+
+### BMO Proposal
+
+Add a new optional section in the BareMetalHost spec containing details of the
+non-bootable iso image, example :
+
+#### Design spec
+
+```yaml
+  spec:
+    dataImage:
+      url: http://1.2.3.4/image.iso
+```
+
+We may consider adding support for HTTP credentials and TLS settings in the future
+but it's outside the scope of this proposal.
+
+**Note** : As part of this proposal, the dataImage will be attached after the node
+has been Provisioned or ExternallyProvisioned.
+
+### Ironic Support
+
+The [proposal](https://bugs.launchpad.net/ironic/+bug/2033288) for exposing this functionaility in Ironic has been
+accepted by the upstream Ironic community.
+
+The [Ironic implementation](https://review.opendev.org/c/openstack/ironic/+/894918) based on the above proposalon has been completed.
+
+Other than that, we just need the implementation for driver specific support,
+which in this proposal is Redfish.
+
+## Design Details
+
+Add an optional section `DataImage` to `BareMetalHostSpec`, with sub-field
+`Url`.
+
+The ISO will be attached when the `BareMetalHost` object is edited to add
+the `DataImage` spec triggering a reconciliation. In case a reboot is requested
+at the same time using `RebootAnnotation`, the ISO will be attached first.
+
+The `BareMetalHostStatus` will cache the image details, which will inform
+us if the attachment succeeded and will also be used for detachment, either
+when the attachment fails and we retry or when the user explicitly requests
+detachment by removing the `DataImage` spec from `BareMetalHost` object.
+
+A flag will be used at the webhook level to validate if a given driver supports the
+non-bootable ISO feature. Since the support for attaching an ISO as virtual media is
+equivalent for both `PreprovisioningISOImage` and `NonBootableISOImage`, we plan to
+rename the existing flag for `PreprovisioningISOImage` i.e. `SupportsISOPreprovisioningImage`
+to `SupportVirtualMediaISOImage` and use it for both the cases.
+
+### Implementation Details
+
+As part of the design we are assuming that the user will edit the BareMetalHost
+to add the dataImage spec, and might also request reboot via rebootAnnotation
+or power state change via `online` field.
+
+We want to attach the dataImage first, so handling the attachment/detachment of dataImage
+inside the Reconcile function, after reconciling host data and updating hardware details ([code reference](https://github.com/metal3-io/baremetal-operator/blob/1bb45eef449c942711b1c0937ecff2b10a326eb3/controllers/metal3.io/baremetalhost_controller.go#L152) ), makes sense since
```

Comment:
This doesn't make sense to me :)
Doing it outside of the state machine means it can't take into account any of the other state. e.g. is it valid to do this in the middle of... provisioning? inspecting? I suspect not.

Handling it in `actionManageSteadyState()` would make sense to me.

Reply:
Done